### ISSUE TITLE
Is there a nicer way to see vtable's function calls directly in the decompiler?

### ISSUE THREAD

**Issue by aldelaro5** [NationalSecurityAgency/ghidra#516]
When I am dealing with a C++ binary, it's expected that I will have to deal with vtables and that the binary will call functions by accessing the instance's vtables. Defining the vtable doesn't seem too difficult as I can type the data to an array of func* and it shows perfectly in the listing. I can then retype the appropriate field in the class's structure to func** so the decompiler knows it is an array of func*.

The problem is this is the best I can get in the decompiler (this is an example I made up to illustrate my point):
```
(*this->field_0x0[6])(local_20);
```
Now, at least it does tell me that it is calling the 7th function in the vtable, but the problem is that I would like the decompiler to be able to infer WHICH function it is called, but it can't because it cannot know that this field is a vtable so it won't really change once assigned which would have been done before. I haven't found an option to set a structure's field as a constant that will never change so I am forced to manually check the vtable in the listing to figure out what is index 6 in the table.

The only other solution I seem to have found is to create a new structure with all the entries being there, but not only it will JUST say the function name and not show the actual reference, it will be completely separate from the class structure which is incredibly inconvenient to create for all vtables as I am dealing with hundreds of classes (my particula binary has debugging information).

My question: Is there a better way to deal with this and if there isn't, would it be possible to fix this? It seems to be a huge inconvenience to check the table every time I see an indirect function call.

**Comment by BhaaLseN** [comment]
I can get to 
```(*this->vtable.MyClass::MyMethod)(this, param1, param2)```
by typing/naming the members of a vtable structure appropriately and then typing the vtable pointer of the actual structure with it.

So basically: `MyClass` with a `vtable` member of type `vtable-MyClass*` and `vtable-MyClass` with a `MyClass::MyMethod` member (name can be anything, I chose to put the class name so it shows up for derived classes) of type `void __thiscall MyMethod(MyClass* this, int param1, int param2)`

It works, but is still kinda clunky in syntax. Helps readability more than `this->vtable[6]` though.

But yeah, it would be nice if there was a way to help this vtable reconstruction; be it structure definition based on the vtables from metadata, naming from the referenced methods or just a vtable field that lets you edit this in the structure (or class!) somewhere.

**Comment by rokups** [comment]
@BhaaLseN how is this done? I can see classes in symbol tree, but when i try to retype a variable - it expects type from "data type" tab. Classes are not there of course. Is there some extra step to create a data type from a class in symbol tree?

**Comment by BhaaLseN** [comment]
Yeah, this needs to be done _by hand_ at this point; which is why I also put my 👍 for better support.

But as mentioned, right now you need _two_ structs; your actual object/class/structure (which was probably auto-created) and a second one for the vtable (which you have to do by hand); then type the vtable member of the first one with a pointer to the second one.

**Comment by mattypiper** [comment]
ghidra vtable guide: http://hwreblog.com/projects/ghidra.html

**Comment by dvdkon** [comment]
I made [a script](https://gitlab.com/dvdkon/ghidra_scripts/blob/master/ProcessVTables.py) for detecting vtables and creating datatypes for them. It does this by looking for "typeinfo name" mangled strings, finding typeinfos by looking at references and then finding vtables the same way. I haven't tested it on more than two binaries, so it's probably got some bugs, but it should work on any Itanium ABI binary (GCC, clang and some others).

Don't worry if you run it and get a lot of "ERR" messages, those might just be artifacts of typename-like strings in the binary or classes where the tool can't get all the necessary data.

It doesn't deal with inheriting from a multiple inheritance class yet, but I'll hopefully fix that later.

Unfortunately, this script can't currently represent the type hierarchy faithfully (i.e. how `gdb`'s `print` does it). This script creates multiple `vptr`s in a struct at particular offsets, but what's really happening is that entire objects (of parent classes) are embedded in the child object. This is all fine and could be implemented easily, but the first embedded object's vptr is also the whole object's vptr, so the pointed-to vtable has extra function pointers appended at the end. This would require something like the types being parametric over their vtable type (or just "hardcoded" support for C++ inheritance) in Ghidra.

**Comment by dvdkon** [comment]
Some people from the CMU have written a [framework](https://insights.sei.cmu.edu/sei_blog/2015/08/the-pharos-framework-binary-static-analysis-of-object-oriented-code.html) for static analysis of object-oriented code, including a tool for recovering classes and methods and a corresponding [Ghidra plugin](https://insights.sei.cmu.edu/sei_blog/2019/07/using-ooanalyzer-to-reverse-engineer-object-oriented-code-with-ghidra.html). It looks much more comprehensive than my script and should be the best way going forward.

**Comment by happydpc** [comment]
> Some people from the CMU have written a [framework](https://insights.sei.cmu.edu/sei_blog/2015/08/the-pharos-framework-binary-static-analysis-of-object-oriented-code.html) for static analysis of object-oriented code, including a tool for recovering classes and methods and a corresponding [Ghidra plugin](https://insights.sei.cmu.edu/sei_blog/2019/07/using-ooanalyzer-to-reverse-engineer-object-oriented-code-with-ghidra.html). It looks much more comprehensive than my script and should be the best way going forward.

But it cannot compile on windows.

**Comment by pabx06** [comment]
been doing like @BhaaLseN said typing function def (names & types) and using them in vtable struct def and applying to vtable data. this work nicely and decompiler are able to infer argument and their types.

However i have like a 100MB binary reversed almost all function def , names and their types but creating manually vtable ... for binary over 100k functions not fun...

At least if only their was was a plugin or script you select the vtable pointer array and hit F2 and it auto create one for you with pointer pointee types that you already reversed !

**Comment by pabx06** [comment]
However creating vtable data structure from pointer to function : do create a structure like : this.

![image](https://user-images.githubusercontent.com/4013370/99806629-94d6bd80-2b3e-11eb-924f-1b2b5f60c221.png)

what is bad: is that functions are already typed correctly. so mutch manual labor for some very low reward 

## Small example
Decided to compile a small example to  see what ghidra's decompiler is able to recover:

![image](https://user-images.githubusercontent.com/4013370/99812796-2a764b00-2b47-11eb-8800-b98b07aa55d4.png)



## 1) inserting func def in the datatype manager
![image](https://user-images.githubusercontent.com/4013370/99811777-e46cb780-2b45-11eb-96b1-1d52b59da1d1.png)
![image](https://user-images.githubusercontent.com/4013370/99811812-f0587980-2b45-11eb-87b4-71060d657671.png)
![image](https://user-images.githubusercontent.com/4013370/99811868-0403e000-2b46-11eb-8ea1-e4960fa54f55.png)
![image](https://user-images.githubusercontent.com/4013370/99811924-154cec80-2b46-11eb-9c6a-38931137542c.png)


## 2) replacing pointer with function def in the vtable and vtable placeholder 
![image](https://user-images.githubusercontent.com/4013370/99812459-c6ec1d80-2b46-11eb-8a23-769dadf5d000.png)

![image](https://user-images.githubusercontent.com/4013370/99812501-d4090c80-2b46-11eb-9e69-f5ae967207b0.png)

![image](https://user-images.githubusercontent.com/4013370/99812545-e420ec00-2b46-11eb-96cf-a95031daaf7e.png)


## Finish product : 
![image](https://user-images.githubusercontent.com/4013370/99813508-0ebf7480-2b48-11eb-90bf-a08a9bec63f1.png)
![image](https://user-images.githubusercontent.com/4013370/99814003-ade46c00-2b48-11eb-87d5-2cc361cbe4dd.png)

a bit disappointed 
- the aes_p->mode  enumeration make the decompiler  choke off . on others sample the enumeration works mostly fine most of the time. is this a bug ? 
- aes_init() : returned data could have been the first field of the struct but it is not !

- main return type and stack variable recovered ok.


here is src,bin,makefile,ghidra project file gzf 
[aes.tar.gz](https://github.com/NationalSecurityAgency/ghidra/files/5574297/aes.tar.gz)

**Comment by dragonmacher** [comment]
@pabx06 That is one heck of a write-up.  That will make it easier to follow your issue.

**Comment by pabx06** [comment]
> That is one heck of a write-up

i cant agree more. typing vtable types so much labor . it would be nice to make a script to pull func-def & name from sected pointer and create a vtable ? imagine 100MB sample to reverse ...

**Comment by brandonros** [comment]
> I made [a script](https://gitlab.com/dvdkon/ghidra_scripts/blob/master/ProcessVTables.py) for detecting vtables and creating datatypes for them. It does this by looking for "typeinfo name" mangled strings, finding typeinfos by looking at references and then finding vtables the same way. I haven't tested it on more than two binaries, so it's probably got some bugs, but it should work on any Itanium ABI binary (GCC, clang and some others).
> 
> Don't worry if you run it and get a lot of "ERR" messages, those might just be artifacts of typename-like strings in the binary or classes where the tool can't get all the necessary data.
> 
> It doesn't deal with inheriting from a multiple inheritance class yet, but I'll hopefully fix that later.
> 
> Unfortunately, this script can't currently represent the type hierarchy faithfully (i.e. how `gdb`'s `print` does it). This script creates multiple `vptr`s in a struct at particular offsets, but what's really happening is that entire objects (of parent classes) are embedded in the child object. This is all fine and could be implemented easily, but the first embedded object's vptr is also the whole object's vptr, so the pointed-to vtable has extra function pointers appended at the end. This would require something like the types being parametric over their vtable type (or just "hardcoded" support for C++ inheritance) in Ghidra.

@dvdkon 

```
ProcessVTables.py> Running...
Traceback (most recent call last):
  File "/Users/brandonros/ghidra_scripts/ProcessVTables.py", line 614, in <module>
    create_typeinfo_vmi_class_type()
  File "/Users/brandonros/ghidra_scripts/ProcessVTables.py", line 403, in create_typeinfo_vmi_class_type
    dt.setFlexibleArrayComponent(
AttributeError: 'ghidra.program.database.data.StructureDB' object has no attribute 'setFlexibleArrayComponent'
ProcessVTables.py> Finished!
```

Incompatible for Ghidra v10?

**Comment by mvf** [comment]
> Incompatible for Ghidra v10?

This seems to work for me. Also adds 64-bit support:

```diff
--- a/ProcessVTables.py
+++ b/ProcessVTables.py
@@ -20,7 +20,7 @@ from collections import namedtuple
 from ghidra.program.util import ProgramMemoryUtil
 from ghidra.program.model.data import (
     StructureDataType, BuiltInDataTypeManager, CategoryPath,
-    FunctionDefinitionDataType, StringDataInstance)
+    FunctionDefinitionDataType, StringDataInstance, ArrayDataType)
 from ghidra.program.model.symbol import SourceType
 from ghidra.program.model.address import Address
 from ghidra.app.util.demangler.gnu import GnuDemanglerNativeProcess
@@ -29,7 +29,7 @@ from ghidra.app.util.demangler.gnu import GnuDemanglerNativeProcess
 SAVE_LOCATION = None

 MAIN_TYPE_CATEGORY_ID = 0 # My guess is it's always 0
-SIZE_T = 4
+SIZE_T = currentProgram.getAddressFactory().getDefaultAddressSpace().getSize() >> 3

 dtm = currentProgram.getDataTypeManager()
 bdtm = BuiltInDataTypeManager.getDataTypeManager()
@@ -400,8 +400,8 @@ def create_typeinfo_vmi_class_type():
     dt.add(charptr, SIZE_T, "__type_name", "")
     dt.add(bdtm.getDataType("/int"), 4, "__flags", "")
     dt.add(bdtm.getDataType("/int"), 4, "__base_count", "")
-    dt.setFlexibleArrayComponent(
-        dtm.getDataType("/__base_class_type_info"), "__base_info", "")
+    dt.add(ArrayDataType(
+        dtm.getDataType("/__base_class_type_info"), 0, -1), "__base_info", "")

 def get_func_str_repr(fsig):
     out = fsig.getPrototypeString()
@@ -454,13 +454,13 @@ def add_type_field_at(dt, cdt, offset, name, desc):
     dtlen = 0 if dt.getNumComponents() == 0 else dt.getLength()
     dt.growStructure(max(0, offset - dtlen))
     j = offset
-    to_delete = []
+    to_delete = set()
     if dt.getNumComponents() > 0:
         while j < offset + clen:
             if j >= dtlen: break
             c = dt.getComponentAt(j)
             j = c.getOffset() + c.getLength()
-            to_delete.append(c.getOrdinal())
+            to_delete.add(c.getOrdinal())
     dt.delete(to_delete)
     dt.insertAtOffset(offset, cdt, clen, name, desc)
```

**Comment by etra0** [comment]
I thought so. It's a bit weird there doesn't exist any way to link both since you can have the `Class` with all the methods listed, the struct with the same name and a vtable struct which you can fill with function names, yet Ghidra can't generate any way to navigate to the functions, but I guess that's the whole point of this issue. Thanks for the reply!

**Comment by brandonros** [comment]
Does their currently exist a way to “export” from decompilation enough to roughly be able to make a .cpp / .h file with the Class and its methods defined and then LoadLibrary / GetProcAddress on the class constructor, then call its vftable methods?

**Comment by NyanCatTW1** [comment]
https://github.com/NyanCatTW1/RedMetaClassAnalyzer/blob/main/RedMetaClassAnalyzer.py

While I and @ChefKissInc were reverse engineering a couple of AMD driver kexts, I created this script that made our experience with vtables much less painful.
It comes with many other features, so you will have to edit the script before it would work.
It will only work on the master branch due to various usages of private functions. It is known to work on Ghidra commit a9f21c738b38c02e3f3e2da78839dafe12a7880

# Features (Newest at bottom)
# Rename metaClass/vtable pointers in __got.
# Find all references to safeMetaCast and retype variables according to the arguments fed.
# Add missing meta structs.
# Add many long fields to ATIController in order to ease the REing of its structure.
# Set up meta/vtable structs to display function name on vtable calls.
# Create vtable stubs in order to ease the REing of its structure.

**Comment by ghidra007** [comment]
There is a prototype/proof of concept script in Ghidra called RecoverClassesFromRTTIScript.java that will figure out the class information if there is RTTI in the program for Windows programs and some gcc programs. it is very rough at this point but you are welcome to try it and see if it works on your binary. If it doesn't, please let us know what the output says and which type of binary you are trying it on. The script figures out class hierarchy and creates class structures and applies them. It figures out which virtual functions belong to which classes and puts them in the correct class and more. Look for the class structures in the data type manager under ClassDataTypes. There is a folder for each class. There are subfolders if a class is in another namespace. If you look in the description of the class structure you will see parent information.  Look in the SymbolTree under Classes for the class members. You can optionally have a class hierarchy graph pop up (edit script to turn on option) or run later using the GraphClassesScript. Eventually, once more class support is available in Ghidra, the script features will be polished and put into main Ghidra. 

We are currently working on a more generic class recovery script for classes without RTTI. It finds vftables, figures out class hierarchy where it can and basically does the same as described above.

**Comment by Wall-AF** [comment]
Is there any chance of making this understand classes built by old Borland C/C++ compilers?

**Comment by ghidra007** [comment]
Are they the ones with the old RTTI format? If so, we have a fix in progress. Once that fix is added to the RTTI analyzer then yes this will work with that.

**Comment by Wall-AF** [comment]
> Are they the ones with the old RTTI format? If so, we have a fix in progress. Once that fix is added to the RTTI analyzer then yes this will work with that.

That I don't know!  But maybe [this reference](https://github.com/mildred/Lysaac/blob/master/doc/boa.utf8.txt) might help you clever folks.

**Comment by ghidra007** [comment]
> > Are they the ones with the old RTTI format? If so, we have a fix in progress. Once that fix is added to the RTTI analyzer then yes this will work with that.
> 
> That I don't know! But maybe [this reference](https://github.com/mildred/Lysaac/blob/master/doc/boa.utf8.txt) might help you clever folks.

Wish we could claim to be clever. Thanks for the nice bit of light reading. :-) Appreciate the pointer to the documentation.

**Comment by 0xBEEEF** [comment]
@ghidra007 
If you' re already on the topic of Borland, it would be great if you also consider Delphi. There is supposed to be some overlap between the RTTI data between C++ and Delphi according to some discussions like [this](https://github.com/crypto2011/IDR/issues/37#issuecomment-359193415) one.

Here you should also find mature analysis methods in [IDR](https://github.com/crypto2011/IDR) that can handle this RTTI data. Maybe just as an idea.

**Comment by ghidra007** [comment]
@0xBEEEF Thanks for the request and links. There are a ton of things on the list for handling various forms of RTTI. Hopefully we can get to them all at some point.

**Comment by markasoftware** [comment]
> We are currently working on a more generic class recovery script for classes without RTTI. It finds vftables, figures out class hierarchy where it can and basically does the same as described above.

Is this on any public branch so we can hack on it?

**Comment by ghidra007** [comment]
> > We are currently working on a more generic class recovery script for classes without RTTI. It finds vftables, figures out class hierarchy where it can and basically does the same as described above.
> 
> Is this on any public branch so we can hack on it?

No sorry. It isn't ready for that yet.

**Comment by ghidra007** [comment]
@vogelfreiheit If the RTTI script doesn't currently work on it, which does have vtable reconstruction in it for some gcc compiled programs, I'm guessing that some changes that are in progress that will make the scirpt work better with gcc compiled programs.

**Comment by ScottJensen18** [comment]
> @ghidra007 If you' re already on the topic of Borland, it would be great if you also consider Delphi. There is supposed to be some overlap between the RTTI data between C++ and Delphi according to some discussions like [this](https://github.com/crypto2011/IDR/issues/37#issuecomment-359193415) one.
> 
> Here you should also find mature analysis methods in [IDR](https://github.com/crypto2011/IDR) that can handle this RTTI data. Maybe just as an idea.

@ghidra007 Also interested in Dephi support!

**Comment by jasper310899** [comment]
Hey, I want to add to this that currently it is not even possible to model vtables faithfully at all or am I wrong? Because imagine you have to model inheritance: childclasses define their own virtual methods for the grandchildren and they collide at the same offset to the vtable start so we can't have names explaining the vfunctions. This is because whith inheritance I always need to define the vptr at the baseclass.

**Comment by ghidra007** [comment]
@jasper310899 The RecoverClassesFromRTTI, a very rough prototype script that needs a lot of cleanup, is an example of how to model vftables at least for Windows binaries. Not sure it is the best way but it works pretty well.  The code is pretty hard to follow so I don't recommend trying to understand it, but if you run it on a windows binary with RTTI you should be able to get a good idea of how it handles things.  Look at structures it creates in the "ClassDataTypes" folder and look at the actual vftables themselves to see the class namespaces the functions are assigned to. It figures out which classes the virtual functions belong to so you can see which are inherited and which are overridden. 

I imagine if there were no virtual methods at all, there would be no vftable and the constructor wouldn't point a virtual function table.

**Comment by jasper310899** [comment]
@ghidra007 I found a perfect solution:
![grafik](https://github.com/NationalSecurityAgency/ghidra/assets/81310031/f156e480-01b6-403b-982a-c67350accd9f)
Using comments you have complete freedom.

**Comment by ghidra007** [comment]
Very true.

**Comment by f-raZ0R** [comment]
> > > We are currently working on a more generic class recovery script for classes without RTTI. It finds vftables, figures out class hierarchy where it can and basically does the same as described above.
> > 
> > 
> > Is this on any public branch so we can hack on it?
> 
> No sorry. It isn't ready for that yet.

How's the progress on that class recovery script going?

**Comment by ghidra007** [comment]
> > > > We are currently working on a more generic class recovery script for classes without RTTI. It finds vftables, figures out class hierarchy where it can and basically does the same as described above.
> > > 
> > > 
> > > Is this on any public branch so we can hack on it?
> > 
> > 
> > No sorry. It isn't ready for that yet.
> 
> How's the progress on that class recovery script going [@ghidra007](https://github.com/ghidra007)?

It's been sidetracked a bit by other things. I hope to get back to it soon.

**Comment by gix** [comment]
After manually adding vtables Ghidra still doesn't seem to use the function signature of the vtable entry at the call sites but continues to use whatever it guessed on initial analysis. I can override the signature manually for each call site to match but that is obviously rather annoying. How do you actually force Ghidra to use the signature of the called function pointer?

**Comment by Torphedo** [comment]
I imported the DirectX 11 headers which have a vtable struct with function tables, and it worked perfectly in Ghidra. If you're at your wits' end, maybe try writing what you want in a C header and importing it.

**Comment by Wall-AF** [comment]
> > > Are they the ones with the old RTTI format? If so, we have a fix in progress. Once that fix is added to the RTTI analyzer then yes this will work with that.
> > 
> > 
> > That I don't know! But maybe [this reference](https://github.com/mildred/Lysaac/blob/master/doc/boa.utf8.txt) might help you clever folks.
> 
> Wish we could claim to be clever. Thanks for the nice bit of light reading. :-) Appreciate the pointer to the documentation.

@ghidra007 is this still on the radar?

**Comment by Lancelotbronner** [comment]
I'm getting this trying to run GCC class recovery:

```
RecoverClassesFromRTTIScript.java> Running...
Checking for missing RTTI information and undefined constructor/destructor functions and creating if possible to find entry point...
Recovering classes using RTTI...
2026-06-02 22:50:15 DEBUG (RTTIGccClassRecoverer) Creating Special Typeinfos  
2026-06-02 22:50:15 DEBUG (RTTIGccClassRecoverer) Creating Special Vtables  
2026-06-02 22:50:16 DEBUG (RTTIGccClassRecoverer) Could not create special vtables  
Error recovering classes
RecoverClassesFromRTTIScript.java> Finished!
```

How can I fix this? I can locate the special vtables and typeinfos manually in Hopper disassembler, it's in extern code.

<img width="307" height="465" alt="Image" src="https://github.com/user-attachments/assets/4cbb9627-2dab-4b37-96f2-e0414a2ddfed" />

<img width="421" height="412" alt="Image" src="https://github.com/user-attachments/assets/0b9c737d-4297-433e-af23-cbaf2cbfb492" />

<img width="1044" height="330" alt="Image" src="https://github.com/user-attachments/assets/de3b3a5d-cb37-437b-861c-d95d48d8e2e6" />

<img width="718" height="502" alt="Image" src="https://github.com/user-attachments/assets/46d6f8de-63ed-400b-846f-3c217c3f6e2a" />

**Comment by Lancelotbronner** [comment]
I managed to get a fix for the C++ script to locate the vtables, macOS binaries have a `_` prefix on all mangled names which confuses the script. I'm not sure if that breaks other things within. [link to my comment in #9243](https://github.com/NationalSecurityAgency/ghidra/issues/9243#issuecomment-4613429601)

The decompiler output hasn't changed, was the C++ script supposed to improve it?

It in fact made it worse, `uVar3 = std::__function::__func<>::__func<>((__func<> *)&app,argc,argv);` used to be `radiant::client::Application::Run(&app,argc,argv);`

```cpp

/* WARNING: Unknown calling convention: __cdecl -- yet parameter storage is locked */
/* lua_main(int, char const**) */

ulong __cdecl lua_main(uint argc,long argv)

{
  undefined *puVar1;
  undefined *puVar2;
  long *this;
  ulong uVar3;
  Application app;
  undefined *local_50;
  ulong argc2;
  long argv2;
  undefined8 local_38;
  long *local_30;
  
  this = operator.new(0x20);
  puVar2 = PTR_vtable_1018ef390;
  puVar1 = PTR_vtable_1018ef390 + 0xc0;
  *this = (long)(PTR_vtable_1018ef390 + 0x38);
  this[1] = (long)puVar1;
  this[2] = (long)(puVar2 + 0x148);
  *(undefined4 *)(this + 3) = 0;
  (**(code **)*this)(this);
  local_38 = 0;
  local_50 = PTR_vtable_1018e4f68 + 0x10;
  argc2 = (ulong)argc;
  argv2 = argv;
  local_30 = this;
  (*(code *)**(undefined8 **)((long)this + *(long *)(*this + -0x38)))
            ((long)this + *(long *)(*this + -0x38));
  uVar3 = CefExecuteProcess((long)&local_50,(long *)&local_30,0);
  if (local_30 != (long *)0x0) {
    (**(code **)(*(long *)((long)local_30 + *(long *)(*local_30 + -0x38)) + 8))
              ((long)local_30 + *(long *)(*local_30 + -0x38));
  }
  if ((int)uVar3 < 0) {
    radiant::client::Application::Application(&app);
    uVar3 = std::__function::__func<>::__func<>((__func<> *)&app,argc,argv);
    radiant::client::Application::~Application(&app);
  }
  (**(code **)(*(long *)((long)this + *(long *)(*this + -0x38)) + 8))
            ((long)this + *(long *)(*this + -0x38));
  return uVar3 & 0xffffffff;
}
```

**Comment by ghidra007** [comment]
@Lancelotbronner Are you seeing a ClassDataTypes folder in the data type manager after running the script? Does it have any subfolders?  Did the script give a nice output with a number of classes found? It doesn't look like a class data structure was applied to your this variable which would have improved your decompiler. So it looks like there are still issues even though the script technically ran. From your image above it looks like there is no space between your vtable symbols. Are they located in your external library and referenced by your main prog? If so, I believe we (analysis not the script) normally create a dummy external block to house the external vtable symbols within the binary importing the libs so that they can at least have a bit of room to lay down the reference to the typeinfo structs inside the vtables. I suspect that might be the issue here. Are you running the script from the program that uses the library or are you running it in the library?

**Comment by Lancelotbronner** [comment]
- Yes, a lot of classes were created, many were even filled (vtables+data), most are just empty placeholder structures
- Just noticed a lot of errors were thrown, also took waaaay longer than a similar size Windows PE executable
- I did bypass a number of checks to force it to run multiple times
- This was run after (multiple) auto-analysis

Attached my modifications of `ghidra_scripts`, the `application.log` section where I ran the C++ script and a [link to the binary](https://github.com/user-attachments/files/28552166/Stonehearth.zip).

[ghidra_scripts.zip](https://github.com/user-attachments/files/28565967/ghidra_scripts.zip)

[application.log](https://github.com/user-attachments/files/28565970/application.log)

**Comment by ghidra007** [comment]
@Lancelotbronner The script is very prototype and is meant to be run after initial analysis and is only meant to be run once. It also hasn't been tested on MacOS. Currently it doesn't do well recovering structures for gcc style RTTI if they are complex inheritance types. There is work in progress that is improving that. I'll take a look and see if I can get your particular issue working regarding the mangled class names. I'll also take a look at what errors were thrown and if they are related to the script or the follow on general analysis that gets kicked off after the script is run.

**Comment by Lancelotbronner** [comment]
Thanks! I'll also see if I can help with that!

**Comment by ghidra007** [comment]
@Lancelotbronner Thanks for figuring out the demangle issue that fixed the error running the script. 

I'm working my way through the verbose DEBUG output. They do look mainly related to the rtti script. The ones that say:
DEBUG 1019255b0: invalid typeinfo - no special classtypeinfo ref'd by baseTypeinfo[1] are because the referenced typeinfo is external. It is on the list to get the script working with externals. 

Many of the other DEBUG statements are highlighting areas where this version possibly needs improving or our understanding of whether to expect a particular related item like VTT table is flawed. 

I'm going to run this in my dev branch where I have made a lot of improvements to see how it does there. 

It is slow but much faster than doing it all by hand :-) 

The windows RTTI does much better than the gcc one. It is on the list to improve it. But that is why the pop up comes up that says be aware this is prototype.  Thanks for your patience while we work towards improving it.

**Comment by ghidra007** [comment]
@Lancelotbronner Using  #9243 to track the vtable __ issue.
